### ISSUE TITLE
Name inconsistency causing bundled executable crashing if path contains none ASCII characters

### ISSUE THREAD

**Issue by JokerQyou** [pyinstaller/pyinstaller#1799]
Hi. I've noticed a very strange phenomenon, which is causing the bundle executable to crash on Windows, before user's code get executed. (The crash occurs at the bootload phase)

Create a script named `test.py`, containing the following content:
```python
print 123
```

Create a `spec` configuration by running `pyi-makespec test.py` in the same folder which contains the script, and the generated spec file will have the following content:
```python
# -*- mode: python -*-

block_cipher = None


a = Analysis(['test.py'],
             pathex=['D:\\Work\\DO\\test'],
             binaries=None,
             datas=None,
             hiddenimports=[],
             hookspath=[],
             runtime_hooks=[],
             excludes=[],
             win_no_prefer_redirects=False,
             win_private_assemblies=False,
             cipher=block_cipher)
pyz = PYZ(a.pure, a.zipped_data,
             cipher=block_cipher)
exe = EXE(pyz,
          a.scripts,
          exclude_binaries=True,
          name='test',
          debug=True,
          strip=False,
          upx=True,
          console=True )
coll = COLLECT(exe,
               a.binaries,
               a.zipfiles,
               a.datas,
               strip=False,
               upx=True,
               name='test')
```

Slightly modify the spec file to let the destination executable to have a different basename with the original script, by changing the content: 
```python
exe = EXE(pyz,
          a.scripts,
          exclude_binaries=True,
          name='test',
          debug=True,
          strip=False,
          upx=True,
          console=True )
```
to this content:
```python
exe = EXE(pyz,
          a.scripts,
          exclude_binaries=True,
          name='ttttt',
          debug=True,
          strip=False,
          upx=True,
          console=True )
```

And build the bundle by running `pyinstaller --clean --noconfirm test.spec`.

Finally, copy the bundle folder (which is the `dist/test` folder in this case) to a Windows machine with a username containing none ASCII characters, let's say a user named `中文用户` on a Windows 7 machine, and put the folder on the user's desktop. So the final path will be `C:\Users\中文用户\Desktop\test`.

Enter the folder `test` and run the executable `ttttt.exe`, and it will crash.

By enabling `debug` flag in the spec file, we can see that the crash happens after the loader printed `Loaded functions from python library`, screenshot attached below.
![image](https://cloud.githubusercontent.com/assets/1465267/12578751/e8d407c8-c45d-11e5-86d1-2c0b6ce9fbfc.png)

This is the basic example to reproduce this problem. My test environment is Windows 7 64bit, and the executable was bundled on a Window 10 machine within a virtualenv with Python 2.7.10 32bit and PyInstaller 3.1.

Further notice: if the name in `EXE` and `COLLECT` and the script name are the same, there will be no problem at all. If two of them are different, things might break (and might not sometimes); If they are totally different with each other, it will almost definitely crash if the path contains Chinese characters.

**Comment by JokerQyou** [comment]
Additional information are as follows:

- I've only tested with Chinese characters.
- The debug log is not the same everytime, I've seen it stops at `sys._MEIP...` something.
- The crashing module isn't always `StackHash_addc`, sometimes it's `ntdll.dll`.
- The exception code is always `c0000005`.

**Comment by htgoebel** [comment]
Thanks for this report. Unicode characters are still problem when using Python 2.7. Can you please try this with Python 3.x?

**Comment by htgoebel** [comment]
Notable issue from the screenshot: `_MEIPASS2 is NULL`

**Comment by JokerQyou** [comment]
Yes I agree that unicode is still a big problem when dealing with Python 2.7. At first I thought that was something in my code is causing the crash, so I tried with `from __future__ import unicode_literals` but ended with no luck. Long after that did I realize this might be something inside PyInstaller.
What makes this really strange is that it won't happen if those names are kept the same.
I'm not sure when I'll have the chance to test with Python 3, I've got all Python 2.7 around me. I'll try some other day.

And speaking of `_MEIPASS2`, what is it and what does it have to do with this problem? For as long as I can remember, the first appearance of this variable is always `NULL` during my debugging.

**Comment by codewarrior0** [comment]
According to the debug output, the crash must occur between lines 374 and 394 of `pyi_pythonlib.c`.

I could not reproduce this error, but your test setup produced a different error ("Could not load frozen modules") which I was able to solve. Would you mind trying the latest `develop` branch head and see if the crash still happens?

**Comment by codewarrior0** [comment]
_MEIPASS2 is used on Linux/OS X when running a onefile executable. The process that extracts the files from the onefile archive will launch a second process to run the actual Python scripts, and _MEIPASS2 is used to pass the extraction folder to the second process.

On Windows, no second process is used, so _MEIPASS2 is always NULL. I believe the only time it would not be NULL on Windows is when using multiprocessing.

**Comment by JokerQyou** [comment]
@codewarrior0 Have you tried with those three names totally different? Sometimes name executable starts with the base name of the script does not crash upon executing. I'll let you know after I give `develop` branch some tests.

**Comment by codewarrior0** [comment]
Yes, I tried with all of the different setups you listed, but could not reproduce the exact crash. I presume the reason is because your Windows is set to Chinese language, which will use a different character set than mine when calling legacy Windows functions.

**Comment by JokerQyou** [comment]
With latest `develop` branch I can't reproduce the problem I originally reported, which is the crash during startup. But now the original test case runs into another problem, as the screenshot shows `parameter error` (or `wrong parameters`):

![image](https://cloud.githubusercontent.com/assets/1465267/12605557/fb9b2adc-c500-11e5-8cf3-38762c0ca8a7.png)

I'm using the following content in `test.spec`:

```python
# -*- mode: python -*-

block_cipher = None


a = Analysis(['test.py'],
             pathex=['D:\\Work\\DO\\assistent'],
             binaries=None,
             datas=None,
             hiddenimports=[],
             hookspath=[],
             runtime_hooks=[],
             excludes=[],
             win_no_prefer_redirects=False,
             win_private_assemblies=False,
             cipher=block_cipher)
pyz = PYZ(a.pure, a.zipped_data,
             cipher=block_cipher)
exe = EXE(pyz,
          a.scripts,
          exclude_binaries=True,
          name='ttttt',
          debug=False,
          strip=False,
          upx=True,
          console=True )
coll = COLLECT(exe,
               a.binaries,
               a.zipfiles,
               a.datas,
               strip=False,
               upx=True,
               name='okkkk')

```

**Comment by codewarrior0** [comment]
Could you try again with `debug=True`? It's not clear why `Failed to get ANSI buffer size` is appearing twice.

**Comment by JokerQyou** [comment]
Ah, I'm very sorry that I forgot to turn debug on in the last test, I just regenerated the `spec` file and changed the names. Here's the debug output:

![image](https://cloud.githubusercontent.com/assets/1465267/12605779/f5b8a4da-c502-11e5-85f9-905dfa00c326.png)

And content of the script `test.py`:

```python
# coding: utf-8
import os
import sys

def main():
    print >> sys.stdout, sys.argv

if __name__ == '__main__':
    sys.exit(main())

```

**Comment by htgoebel** [comment]
Should this go into 3.1.1?

(Please mind setting the milestone for solved issues - it's not only for statistics ;-) You ma even assigne th ticket to yourself ;-)

**Comment by htgoebel** [comment]
I decided to put it into 3.1.1., so I can push that release.

**Comment by codewarrior0** [comment]
Whoops. Yeah, this one should be in 3.1.1.

Looks like I had a brain fart and swapped two variable names. Here's the other part of that fix, which should also go into 3.1.1.

**Comment by codewarrior0** [comment]
I'm still not sure this issue is solved, though. @JokerQyou, mind trying the `develop` branch again?

**Comment by htgoebel** [comment]
@codewarrior0 455fdb5 fail at appveyor, though it is test_threading_module2. Is it ready for 3.1.1?

**Comment by codewarrior0** [comment]
That test looks like it has been failing on Py3 since cf0965585086f728c265b76bd3eb12ae43e5b6ec.

There's another failure in Py2.7 that I'm looking at, but it may be because it's running a Py3-only test on 2.7.

**Comment by codewarrior0** [comment]
Yep, it was running a Py3-test on 2.7 because of a bad pytest marker. Fixed now.

**Comment by JokerQyou** [comment]
I'd like to try it now, but I'm actually in vacation now (2 weeks, for the Chinese New Year), and likely won't have access to a proper Windows environment until February 15th. So, sorry about that... You guys can test this and close it if your tests pass. I'll try the latest code later, and reopen this issue if any further problem occurs.

**Comment by htgoebel** [comment]
Forgot closing the issue

**Comment by JokerQyou** [comment]
I tested this again with latest PyInstaller code from `develop` branch, and things start to become mysterious.

- My test script content:
    
    ```python
    # coding: utf-8
    import os
    import sys

    def main():
        print >> sys.stdout, sys.argv
        return 0

    if __name__ == '__main__':
        sys.exit(main())
    ```
- My spec file content:
    ```python
    # -*- mode: python -*-

    block_cipher = None


    a = Analysis(['test.py'],
                 pathex=['D:\\Work\\DO\\Test'],
                 binaries=None,
                 datas=None,
                 hiddenimports=[],
                 hookspath=[],
                 runtime_hooks=[],
                 excludes=[],
                 win_no_prefer_redirects=False,
                 win_private_assemblies=False,
                 cipher=block_cipher)
    pyz = PYZ(a.pure, a.zipped_data,
                 cipher=block_cipher)
    exe = EXE(pyz,
              a.scripts,
              exclude_binaries=True,
              name='my',
              debug=True,
              strip=False,
              upx=True,
              console=True )
    coll = COLLECT(exe,
                   a.binaries,
                   a.zipfiles,
                   a.datas,
                   strip=False,
                   upx=True,
                   name='test')
    ```
- And I did these things:
    - Run `pyinstaller --clean --noconfirm test.spec` in `D:\Work\DO\Test\`
    - Copy the folder `dist\test` into a Windows 7 machine (Chinese language), put it on a user's (`中文用户`) desktop
    - Enter this folder and open a cmd window, run `my.exe`

I get the following results:
- If I have one or more arguments that is/are ASCII chars, it is ok.
    ![image](https://cloud.githubusercontent.com/assets/1465267/13039064/c5a017e2-d3d5-11e5-904d-d71d53ec2283.png)
- If I have two or more arguments, with the first one being ASCII chars, it is still ok.
    ![image](https://cloud.githubusercontent.com/assets/1465267/13039084/f1c16998-d3d5-11e5-8876-2aa60faa9919.png)
- If I have one single argument which is non-ASCII chars, it will crash sometimes.
    ![image](https://cloud.githubusercontent.com/assets/1465267/13039110/3456b286-d3d6-11e5-9072-26cae3a50158.png)
- If I have more than one arguments, with the first being non-ASCII chars, it will crash sometimes.
    ![image](https://cloud.githubusercontent.com/assets/1465267/13039155/898817c2-d3d6-11e5-8d20-297dd8cc0043.png)

Right now I have no idea how to  trigger the crash in the last situation. I noticed `my.exe 外壳 应用程序扩展 1` would crash every time, however.

**Comment by AndCycle** [comment]
@JokerQyou 
I recently fixed a nasty bug #1889,
you might wanna try again
